dali/kernels/common/block_setup.h

```diff
@@ -0,0 +1,263 @@
+    return uniform_output_size_;
+  }
+
+  ivec<ndim> UniformBlockSize() const {
+    assert(is_uniform_);
+    return uniform_block_size_;
+  }
+
+  bool IsUniformSize() const { return is_uniform_; }
+
+  static inline ivec<ndim> shape2size(const TensorShape<tensor_ndim> &shape) {
+    return shape2vec(skip_dim<channel_dim>(shape));
+  }
+
+ private:
+  std::vector<BlockDesc> blocks_;
+  ivec3 block_dim_{32, 32, 1};
+  ivec3 grid_dim_{1, 1, 1};
+  ivec<ndim> uniform_block_size_, uniform_output_size_;
+  ivec<ndim> default_block_size_;
+  int max_block_elements_ = 1<<16;
+  bool is_uniform_ = false;
+
+  template <int d>
+  void MakeBlocks(BlockDesc blk, ivec<ndim> size, ivec<ndim> block_size,
+                  std::integral_constant<int, d>) {
+    for (int i = 0; i < size[d]; i += block_size[d]) {
+      blk.start[d] = i;
+      blk.end[d] = std::min(i + block_size[d], size[d]);
+      if (d > 0) {
+        constexpr int next_d = d > 0 ? d - 1: d;  // prevent infinite template expansion
+        MakeBlocks(blk, size, block_size, std::integral_constant<int, next_d>());
+      } else {
+        blocks_.push_back(blk);
+      }
+    }
+  }
+
+  void MakeBlocks(int sample_idx, ivec<ndim> size, ivec<ndim> block_size) {
+    BlockDesc blk;
+    blk.sample_idx = sample_idx;
+    MakeBlocks<ndim-1>(blk, size, block_size, {});
+    grid_dim_ = ivec3(blocks_.size(), 1, 1);
+  }
+
+  ivec<ndim> BlockSize(const ivec<ndim> &shape) const {
+    ivec<ndim> ret;
+    for (int i = 0; i < ndim; i++) {
+      switch (i) {
+      case 0:
+      case 1:
+        ret[i] = default_block_size_[i];
+        break;
+      case 2:
+        ret[i] = std::max<int>(1, volume(shape) / max_block_elements_);
+        break;
+      default:
+        ret[i] = 1;
+      }
+    }
+    return ret;
+  }
+
+  void VariableSizeSetup(const TensorListShape<tensor_ndim> &output_shape) {
+    for (int i = 0; i < output_shape.num_samples(); i++) {
+      ivec<ndim> size = shape2size(output_shape[i]);
+      ivec<ndim> block_size = BlockSize(size);
+      MakeBlocks(i, size, block_size);
+    }
+  }
+
+  void UniformSizeSetup(const TensorListShape<tensor_ndim> &output_shape) {
+    if (output_shape.empty())
+      return;
+    uniform_output_size_ = shape2size(output_shape[0]);
+    // Get the rough estimate of block size
+    uniform_block_size_ = BlockSize(uniform_output_size_);
+
+    // Make the blocks as evenly distributed as possible over the target area,
+    // but maintain alignment to CUDA block dim.
+    for (int i = 0; i < 2; i++) {  // only XY dimensions
+      int blocks_in_axis = div_ceil(uniform_output_size_[i], uniform_block_size_[i]);
+      int even_block_size = div_ceil(uniform_output_size_[i], blocks_in_axis);
+
+      // a note on div_ceil + mul combo:
+      // can't use align_up, because block_dim_ does not need to be a power of 2
+      uniform_block_size_[i] = div_ceil(even_block_size, block_dim_[i]) * block_dim_[i];
+    }
+
+    grid_dim_ = {
```

Comment:
Is it intended to work with more than HWC dimensions? Because you cover here only the first two dimensions, the third dim in grid is just a sample number. What if we have 3D+channels data, it won't take into account iterating over third dim. How should that case be handled?

Reply:
The intent was that other dimensions are iterated by threads. It's what I do in warp, although arguably this is not a very generic solution.

Reply:
Made it explicit; tested.